src/strkey.rs

```diff
@@ -111,17 +127,167 @@ impl FromStr for StrkeyPrivateKeyEd25519 {
     }
 }
 
+#[derive(Clone, Copy, Hash, PartialEq, Eq, PartialOrd, Ord, Debug)]
+pub struct StrkeyMuxedAccountEd25519 {
+    pub ed25519: [u8; 32],
+    pub id: u64,
+}
+
+impl StrkeyMuxedAccountEd25519 {
+    pub fn to_string(&self) -> String {
+        let mut payload: [u8; 40] = [0; 40];
+        let (ed25519, id) = payload.split_at_mut(32);
+        ed25519.copy_from_slice(&self.ed25519);
+        id.copy_from_slice(&self.id.to_be_bytes());
+        encode(version::MUXED_ACCOUNT_ED25519, &payload)
+    }
+
+    fn from_payload(payload: &[u8]) -> Result<Self, DecodeError> {
+        let (ed25519, id) = payload.split_at(32);
+        Ok(Self {
+            ed25519: ed25519.try_into().map_err(|_| DecodeError::Invalid)?,
+            id: u64::from_be_bytes(id.try_into().map_err(|_| DecodeError::Invalid)?),
+        })
+    }
+
+    pub fn from_string(s: &str) -> Result<Self, DecodeError> {
+        let (ver, payload) = decode(s)?;
+        match ver {
+            version::MUXED_ACCOUNT_ED25519 => Self::from_payload(&payload),
+            _ => Err(DecodeError::Invalid),
+        }
+    }
+}
+
+#[derive(Clone, Copy, Hash, PartialEq, Eq, PartialOrd, Ord, Debug)]
+pub struct StrkeyPreAuthTx(pub [u8; 32]);
+
+impl StrkeyPreAuthTx {
+    pub fn to_string(&self) -> String {
+        encode(version::PRE_AUTH_TX, &self.0)
+    }
+
+    fn from_payload(payload: &[u8]) -> Result<Self, DecodeError> {
+        Ok(Self(payload.try_into().map_err(|_| DecodeError::Invalid)?))
+    }
+
+    pub fn from_string(s: &str) -> Result<Self, DecodeError> {
+        let (ver, payload) = decode(s)?;
+        match ver {
+            version::PRE_AUTH_TX => Self::from_payload(&payload),
+            _ => Err(DecodeError::Invalid),
+        }
+    }
+}
+
+#[derive(Clone, Copy, Hash, PartialEq, Eq, PartialOrd, Ord, Debug)]
+pub struct StrkeyHashX(pub [u8; 32]);
+
+impl StrkeyHashX {
+    pub fn to_string(&self) -> String {
+        encode(version::HASH_X, &self.0)
+    }
+
+    fn from_payload(payload: &[u8]) -> Result<Self, DecodeError> {
+        Ok(Self(payload.try_into().map_err(|_| DecodeError::Invalid)?))
+    }
+
+    pub fn from_string(s: &str) -> Result<Self, DecodeError> {
+        let (ver, payload) = decode(s)?;
+        match ver {
+            version::HASH_X => Self::from_payload(&payload),
+            _ => Err(DecodeError::Invalid),
+        }
+    }
+}
+
+/// Stores a signed payload ed25519 signer.
+///
+/// The payload must not have a size larger than u32::MAX.
+#[derive(Clone, Hash, PartialEq, Eq, PartialOrd, Ord, Debug)]
+pub struct StrkeySignedPayloadEd25519 {
+    pub ed25519: [u8; 32],
+    pub payload: Vec<u8>,
```

Comment:
Hi @leighmcculloch, I want to check with you if the `payload` can be empty.

Reply:
I was wrong about this. The network considers the Strkey malformed if empty. https://github.com/stellar/js-stellar-base/issues/556#issuecomment-1278349601

Reply:
Although that constraint is defined in the protocol, while the Strkey doesn't define that limitation. https://github.com/stellar/stellar-protocol/blob/master/ecosystem/sep-0023.md

Reply:
Thanks, it's clear now, the correction is in c702bc99603e35770b28224406a6f251f5b6f566

Reply:
The Go SDK allows zero length it seems. 

I think it's debatable what SDKs should implement. While the Strkey spec doesn't say zero length is disallowed, since the protocol considers a zero length payload as malformed, it doesn't seem very useful to support them in sdk's. 

It probably will have little consequence.

Reply:
Yes, whatever implementation probably won't have much impact on most users, but we need to decide how to implement it in the Rust SDK, and I personally would prefer not to allow it to be set to empty since the Signed Payload has no purpose other than to be submitted to stellar-core? And stellar-core will treat it as **malformed**.

Also, I don't think the SDK should ever construct a [malformed transaction](https://developers.stellar.org/api/errors/http-status-codes/horizon-specific/transaction-malformed/).

Reply:
I agree. I think you're correct. There is little to no gain to supporting zero-length payload. And there is harm to supporting zero-length payload.